content/courses/advanced-ada/chapters/subprograms.rst

```diff
@@ -3,6 +3,287 @@ Subprograms
++--------------------+--------------------------------------------------------+
+| Binary adding      | :ada:`+`, :ada:`-`, :ada:`&`                           |
++--------------------+--------------------------------------------------------+
+| Multiplying        | :ada:`*`, :ada:`/`, :ada:`mod`, :ada:`rem`             |
++--------------------+--------------------------------------------------------+
+| Highest precedence | :ada:`**`, :ada:`abs`, :ada:`not`                      |
++--------------------+--------------------------------------------------------+
+
+.. admonition:: In the Ada Reference Manual
+
+    - `4.5 Operators and Expression Evaluation <http://www.ada-auth.org/standards/12rm/html/RM-4-5.html>`_
+
+User-defined operators
+~~~~~~~~~~~~~~~~~~~~~~
+
+For non-scalar types, not all operators are defined. For example, it wouldn't
+make sense to expect a compiler to include an addition operator for a record
+type with multiple components. An exception to this rule are the
+equality and inequality operators (:ada:`=` and :ada:`/=`), which are defined
+for any type (be it scalar, record types, and array types).
+
+For array types, the concatenation operator (:ada:`&`) is a primitive operator:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Subprograms.Integer_Arrays_Concat
+
+    package Integer_Arrays is
+
+       type Integer_Array is array (Positive range <>) of Integer;
+
+    end Integer_Arrays;
+
+    with Ada.Text_IO;    use Ada.Text_IO;
+    with Integer_Arrays; use Integer_Arrays;
+
+    procedure Show_Array_Concatenation is
+       A, B : Integer_Array (1 .. 5);
+       R : Integer_Array (1 .. 10);
+    begin
+       A := (1 & 2 & 3 & 4 & 5);
+       B := (6 & 7 & 8 & 9 & 10);
+       R := A & B;
+
+       for E of R loop
+          Put (E'Image & ' ');
+       end loop;
+       New_Line;
+    end Show_Array_Concatenation;
+
+In this example, we're using the primitive :ada:`&` operator to concatenate the
+:ada:`A` and :ada:`B` arrays in the assignment to :ada:`R`. Similarly, we're
+concatenating individual components (integer values) to create an aggregate
+that we assign to :ada:`A` and :ada:`B`.
+
+In contrast to this, the addition operator is not available for arrays:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Subprograms.Integer_Arrays_Addition
+    :class: ada-expect-compile-error
+
+    package Integer_Arrays is
+
+       type Integer_Array is array (Positive range <>) of Integer;
+
+    end Integer_Arrays;
+
+    with Ada.Text_IO;    use Ada.Text_IO;
+    with Integer_Arrays; use Integer_Arrays;
+
+    procedure Show_Array_Addition is
+       A, B, R : Integer_Array (1 .. 5);
+    begin
+       A := (1 & 2 & 3 & 4 & 5);
+       B := (6 & 7 & 8 & 9 & 10);
+       R := A + B;
+
+       for E of R loop
+          Put (E'Image & ' ');
+       end loop;
+       New_Line;
+
+    end Show_Array_Addition;
+
+We can, however, define *custom* operators for any type. For example, we can
```

Comment:
This needs some precision: we can define our own operator functions, but the function names (the "designators")  we use must be one of those that are already defined as operator symbols by the language. In other words we cannot make up custom operators using new symbols. Section 4.5 of the Ada RM lists the possible operator symbols available to us.

Reply:
Agreed! I added a paragraph explaining that (and used some of what you said above).